tokio/src/sync/semaphore.rs

```diff
@@ -102,6 +102,78 @@ use std::sync::Arc;
 /// }
 /// ```
 ///
+/// Implement a simple token bucket for rate limiting
+///
+/// Many applications and systems have constraints on the rate at which certain
+/// operations should occur. Exceeding this rate can result in suboptimal
+/// performance or even errors.
+///
+/// The provided example uses a `TokenBucket`, implemented using a semaphore, that
+/// limits operations to a specific rate. The token bucket will be refilled gradually.
+/// When the rate is exceeded, the `acquire` method will await until a token is available.
```

Comment:
We should mention two things here:

 * Token buckets allow short bursts that are faster than the allowed rate. In fact, this is part of the point of token buckets.
 * This implementation is suboptimal when the rate is large, because it consumes a lot of cpu constantly looping and sleeping.